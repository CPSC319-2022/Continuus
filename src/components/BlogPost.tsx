import { MenuIcon } from "~/icons/Menu";

interface BlogPostProps {
  id: number;
  name: string;
  lastUpdated: string;
  title: string;
  content: string;
  imageUrl: string;
  comments: number;
}

export const BlogPost: React.FC<BlogPostProps> = ({
  id,
  name,
  lastUpdated,
  title,
  content,
  imageUrl,
  comments,
}) => {
  return (
    <div className="bg-base-150 card w-full rounded-md shadow-md shadow-slate-300">
      <div className="card-body m-[-15px]">
        <div className="mb-3 flex w-full justify-between">
          <div className="flex">
            <div className="avatar self-center">
              <div className="h-10 w-10 rounded-full">
                <img src={imageUrl} alt="avatar" />
              </div>
            </div>
            <div className="ml-3">
              <p className="text-lg font-bold">{name}</p>
              <p className="text-sm text-slate-400">{lastUpdated}</p>
            </div>
          </div>
          <div className="self-center">
            <div className="dropdown-left dropdown rounded-md shadow-slate-300">
              <button>
                <MenuIcon />
              </button>
              <ul
                tabIndex={0}
                className="dropdown-content menu rounded-box w-52 bg-base-100 p-2 shadow"
              >
                <li>
                  <a>Edit</a>
                </li>
                <li>
                  <a>Delete</a>
                </li>
              </ul>
            </div>
          </div>
        </div>
        <p className="mb-3 text-xl font-bold">{title}</p>
        <p>{content.length > 500 ? `${content.slice(0, 499)}...` : content}</p>
        <div className="self-end ">
          <label
            htmlFor={`modal-${id}`}
            className="btn-link text-highlight-green no-underline hover:cursor-pointer"
          >
            {comments} Comments
          </label>
        </div>
      </div>
    </div>
  );
};
