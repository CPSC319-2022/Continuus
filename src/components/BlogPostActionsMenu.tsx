import { MenuIcon } from "~/icons/Menu";
import { UpdateBlogPostWidget } from "~/components/update-blog-post-widget/UpdateBlogPostWidget";
import React from "react";
import { DeleteBlogPostWidget } from "~/components/delete-blog-post-widget/DeleteBlogPostWidget";

interface BlogPostActionsProps {
  id: string;
  title: string;
  content: string;
  isAuthor: boolean;
}

export const BlogPostActionsMenu: React.FC<BlogPostActionsProps> = ({id, title, content, isAuthor}) => {
  return (
    <div className="self-center">
      <div className="dropdown-left dropdown rounded-md shadow-slate-300">
        <button>
          <MenuIcon />
        </button>
        <ul
          tabIndex={0}
          className="dropdown-content menu rounded-box w-52 bg-base-100 p-2 shadow"
        >
          {
            isAuthor &&
              <li>
                  <UpdateBlogPostWidget id={id} title={title} content={content} />
              </li>
          }

          <li>
            <DeleteBlogPostWidget id={id} />
          </li>
        </ul>
      </div>
    </div>
  );
}