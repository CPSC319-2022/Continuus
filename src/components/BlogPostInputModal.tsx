import type { PropsWithChildren } from 'react';
import React from 'react';
import Modal from 'react-modal';

{process.env.NODE_ENV !== 'test' ? Modal.setAppElement("#__next") : Modal.setAppElement("body")}

export const BlogPostInputModal: React.FC<PropsWithChildren<{
    action: string,
    isOpen: boolean,
    closeDisabled?: boolean,
    onRequestClose: () => void,
}>> = ({
    children,
    action,
    isOpen,
    closeDisabled,
    onRequestClose,
}) => {
        return (
            <Modal
                isOpen={isOpen}
                onRequestClose={() => closeDisabled || onRequestClose()}
                contentLabel="Blog Post Input"
                overlayClassName="fixed inset-0 z-20 bg-black/75"
                ariaHideApp={process.env.NODE_ENV !== 'test'}
                className="bg-white absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 w-11/12 rounded-md md:w-1/2 z-40"
            >
                <div className="w-full h-12 flex flex-row justify-between items-center border-b border-b-gray-200 pl-4 pr-4">
                    <div className="font-semibold">{action} Blog Post</div>
                    {
                        closeDisabled || <button
                            className="rounded-full after:content-['\2715'] hover:text-emerald-400 transition-colors"
                            onClick={() => closeDisabled || onRequestClose()}
                        />
                    }
                </div>
                {children}
            </Modal>
        );
    }