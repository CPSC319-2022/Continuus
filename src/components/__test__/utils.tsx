import React, { type PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import { configureStore, type PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";

import type { AppStore, RootState } from "~/redux/store";
import { reducer } from "~/redux/slices";
// As a basic setup, import your same slice reducers

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({ reducer: reducer, preloadedState }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  // eslint-disable-next-line @typescript-eslint/ban-types
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
