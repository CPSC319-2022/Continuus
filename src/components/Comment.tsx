import { userPathToProfile } from "~/utils/profile";
import { timeAgo } from "~/utils/time";
import { ProfileName } from "./ProfileName";
import { ProfilePicture } from "./ProfilePicture";
import React, { useEffect, useState } from "react";
import { api } from "~/utils/api";
import { useSession } from "next-auth/react";
import { isAuthed, isAuthor } from "~/components/util";
import { Spinner } from "~/components/Spinner";
import { MenuIcon } from "~/icons/Menu";

export interface CommentProps {
  dateAdded: Date;
  comment: string;
  name?: string | null;
  userId: string;
  imgUrl?: string | null;
  commentId: string;
  dateUpdated: Date;
}

export const Comment: React.FC<CommentProps> = ({
  dateAdded,
  comment,
  name,
  userId,
  imgUrl,
  commentId,
  dateUpdated,
}) => {
  const utils = api.useContext();
  const { status } = useSession();
  const [editInput, setEditInput] = useState<string>(comment);
  const [isEditing, setIsEditing] = useState(false);
  const [isDeletePending, setIsDeletePending] = useState(false);
  const [isDeleted, setIsDeleted] = useState(false);
  const currUser = api.user.currentUser.useQuery();

  const updateCommentMutation = api.comment.update.useMutation({
    onSuccess: () =>
      Promise.all([
        utils.comment.invalidate(),
        utils.blogPost.getOne.invalidate(),
        utils.blogPost.get.invalidate(),
      ]),
  });

  const deleteCommentMutation = api.comment.delete.useMutation({
    onSuccess: () =>
      Promise.all([
        utils.blogPost.getOne.invalidate(),
        utils.blogPost.get.invalidate(),
      ]),
  });

  useEffect(() => {
    if (updateCommentMutation.isSuccess) setIsEditing(false);
  }, [updateCommentMutation.isSuccess]);

  useEffect(() => {
    if (deleteCommentMutation.isSuccess) {
      setIsDeleted(true);
      setIsDeletePending(false);
    }
  }, [deleteCommentMutation.isSuccess]);

  const fetchedComment = api.comment.getUnique.useQuery({
    where: {
      id: commentId,
    },
  });

  if (fetchedComment.data !== undefined) comment = fetchedComment.data.content;

  const shouldSeeCommentActions = () => {
    return isAuthed(status) && isAuthor(currUser.data, userId);
  };

  const handleInputChange = (event: {
    target: { value: React.SetStateAction<string> };
  }) => {
    setEditInput(event.target.value);
  };

  const handleUpdateCancel = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.preventDefault();
    setIsEditing(false);
    setEditInput(comment);
  };

  const handleCommentUpdate = () => {
    if (editInput) {
      updateCommentMutation.mutate({
        where: {
          id: commentId,
        },
        data: {
          content: editInput,
        },
      });
    }
  };

  const handleCommentDelete = () => {
    deleteCommentMutation.mutate({
      where: {
        id: commentId,
      },
    });
  };

  return (
    <>
      {isDeletePending ? (
        <div className="mb-4">
          <button
            className="absolute right-12 rounded-full transition-colors transition-all after:content-['\2715'] hover:scale-125 hover:text-highlight-red"
            onClick={(event) => {
              event.preventDefault();
              setIsDeletePending(false);
            }}
          ></button>
          <div className="flex h-20 w-full items-center justify-center">
            <div className="text-lg">
              Are you sure you want to delete this comment?
            </div>
          </div>
          <div className="flex justify-end border-b pb-3">
            <button
              className={
                "h-10 w-32 rounded-md border bg-highlight-red text-center text-white transition-all hover:cursor-pointer hover:bg-gray-700"
              }
              onClick={() => handleCommentDelete()}
            >
              Delete
            </button>
          </div>
        </div>
      ) : (
        !isDeleted && (
          <div className="relative m-3 rounded-md border-b p-4">
            <div className="flex">
              <div className="avatar self-center transition-all hover:scale-110">
                <ProfilePicture
                  size={2.5}
                  redirectLink={userPathToProfile(userId)}
                  imgUrl={imgUrl}
                />
              </div>
              <div className="ml-3">
                <ProfileName name={name || ""} userId={userId} />
                <p className="tex-gray-400 text-sm">
                  {timeAgo(dateAdded)}
                  {dateAdded.getTime() !== dateUpdated.getTime()
                    ? ` (updated ${timeAgo(dateUpdated)})`
                    : ""}
                </p>
              </div>
              {shouldSeeCommentActions() && (
                <div className="absolute right-4 mt-[-12px] self-center">
                  <div className="dropdown-left dropdown rounded-md">
                    <button
                      className="transition-all hover:scale-125"
                      data-testid="comment-actions-button"
                    >
                      <MenuIcon />
                    </button>
                    <ul
                      tabIndex={0}
                      className="dropdown-content menu mt-8 mr-[-16px] w-52 rounded-md border bg-white shadow"
                    >
                      <li
                        data-testid="edit-comment"
                        className="p-3 text-start transition-all hover:cursor-pointer hover:bg-gray-100"
                        onClick={() => setIsEditing(true)}
                      >
                        Edit
                      </li>
                      <li
                        data-testid="delete-comment"
                        className="p-3 text-start transition-all hover:cursor-pointer hover:bg-gray-100"
                        onClick={() => setIsDeletePending(true)}
                      >
                        Delete
                      </li>
                    </ul>
                  </div>
                </div>
              )}
            </div>
            {updateCommentMutation.isLoading ? (
              <div className="flex w-full items-center justify-center">
                <Spinner size={2} />
              </div>
            ) : isEditing ? (
              <div className="mt-3 p-2" data-testid="edit-comment-div">
                <input
                  autoFocus
                  type="text"
                  placeholder="Can't be empty!"
                  value={editInput}
                  onChange={handleInputChange}
                  className="input-bordered mb-3 w-full rounded-md border p-3"
                />
                <div className="flex justify-end gap-2">
                  <button
                    className={`h-10 w-32 rounded-md ${
                      !editInput
                        ? "input-disabled cursor-not-allowed bg-gray-200 text-black"
                        : "border-highlight-green bg-highlight-green transition-all hover:bg-gray-700 hover:text-highlight-green"
                    }`}
                    onClick={() => handleCommentUpdate()}
                    disabled={!editInput ? true : false}
                  >
                    Save
                  </button>
                  <button
                    className="h-10 w-32 rounded-md border bg-white text-center text-black transition-all hover:cursor-pointer hover:bg-highlight-red hover:text-white"
                    onClick={(event) => handleUpdateCancel(event)}
                  >
                    Cancel
                  </button>
                </div>
              </div>
            ) : (
              <div className="break-words">
                <p className="mt-4 w-full">{comment}</p>
              </div>
            )}
          </div>
        )
      )}
    </>
  );
};
