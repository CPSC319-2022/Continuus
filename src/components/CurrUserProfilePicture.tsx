import { api } from "~/utils/api";
import { Spinner } from "./Spinner";

/**
 * @prop size: Width/Height of the Profile Picture in REM 
 * @returns 
 */
export const CurrUserProfilePicture: React.FC<{ size: number }> = ({ size }) => {
    const currUser = api.user.currentUser.useQuery();

    if (currUser.isLoading) {
        return <div className="aspect-square rounded-full" style={{ width: `${size}rem` }}>
            <Spinner size={size} />
        </div>
    }

    return (
        <img src={currUser.data?.image || "https://i.stack.imgur.com/34AD2.jpg"} alt="avatar" className="aspect-square rounded-full" style={{ width: `${size}rem` }} />

    )
}