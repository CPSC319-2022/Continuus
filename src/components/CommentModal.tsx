import { Comment } from "./Comment";
import React, { useState } from "react";
import ReactMarkdown from "react-markdown";
import remarkGfm from "remark-gfm";
import remarkSlug from "remark-slug";
import type { Comment as CommentType, User } from "@prisma/client";
import { timeAgo } from "~/utils/time";
import { ProfilePicture } from "./ProfilePicture";
import { BlogPostActionsMenu } from "~/components/BlogPostActionsMenu";
import { api } from "~/utils/api";
import { useSession } from "next-auth/react";
import { shouldSeeActions, isAuthor } from "~/components/util";
import {userPathToProfile} from "~/utils/profile";

type CommentEntry = CommentType & { user: User };

export interface CommentModalProps {
  id: string;
  title: string;
  poster: string;
  lastUpdated: Date;
  createdAt: Date;
  post: string;
  comments: CommentEntry[];
  content: string;
  authorId: string;
  imgUrl?: string | null;
}

export const CommentModal: React.FC<CommentModalProps> = ({
  id,
  title,
  poster,
  lastUpdated,
  createdAt,
  post,
  comments,
  content,
  authorId,
  imgUrl,
}) => {
  const [input, setInput] = useState<string>("");

  const utils = api.useContext();

  const currUser = api.user.currentUser.useQuery();

  const { status } = useSession();

  const createCommentMutation = api.comment.create.useMutation({
    onSuccess() {
        return Promise.all([utils.blogPost.get.invalidate(),
                            utils.comment.count.invalidate()]);
    },
  });

  const userId = currUser?.data?.id;

  const handleCommentChange = (event: {
    target: { value: React.SetStateAction<string> };
  }) => {
    setInput(event.target.value);
  };

  const handlePostButtonClick = () => {
    createCommentMutation.mutate({
      data: {
        userId: userId!,
        blogPostId: id,
        content: input,
      },
    });

    setInput("");
  };

  return (
    <>
      <input type="checkbox" id={`modal-${id}`} className="modal-toggle" />
      <label htmlFor={`modal-${id}`} className="modal z-10 cursor-pointer">
        <label
          className="card-body modal-box relative m-[-10px] w-11/12 max-w-5xl rounded-md"
          htmlFor=""
        >
          <div className="mb-3 flex w-full justify-between">
            <div className="flex">
              <div className="avatar self-center">
                  <ProfilePicture 
                      size={2.5}
                      redirectLink={userPathToProfile(authorId)} 
                      imgUrl={imgUrl}
                  />
              </div>
              <div className="ml-3">
                <p className="text-lg font-bold">{poster}</p>
                <p className="text-sm text-slate-400">{`${timeAgo(createdAt)}${
                  createdAt.getTime() !== lastUpdated.getTime()
                    ? ` (updated ${timeAgo(lastUpdated)})`
                    : ""
                }`}</p>
              </div>
            </div>
            {
              shouldSeeActions(status, currUser.data, authorId) && (
              <BlogPostActionsMenu
                id={id}
                title={title}
                content={content}
                isAuthor={isAuthor(currUser.data, authorId)}
              />
            )}
          </div>
          <p className="mb-3 text-xl font-bold">{title}</p>
          <div className="prose max-w-none ">
            <ReactMarkdown remarkPlugins={[remarkGfm, remarkSlug]}>
              {post}
            </ReactMarkdown>
          </div>
          <div className="self-end">
            <p className="btn-link text-highlight-green no-underline hover:no-underline">
              {comments.length} Comments
            </p>
          </div>
          <div>
            {comments.map(
              ({
                id: commentId,
                content: comment,
                createdAt: dateAdded,
                user: { name: authorName, id: commenterId, image},
              }) => (
                <Comment
                  key={`${commentId}`}
                  dateAdded={dateAdded}
                  comment={comment}
                  name={authorName}
                  userId={commenterId}
                  imgUrl={image}
                />
              )
            )}
          </div>
          <div className="m-[2rem] mx-0 mb-0">
            <div
              className={`${userId ? "" : "tooltip tooltip-info"} w-full`}
              data-tip="Sign in to comment"
            >
              <input
                type="text"
                placeholder="Write a new comment"
                value={input}
                onChange={handleCommentChange}
                readOnly={userId ? false : true}
                disabled={userId ? false : true}
                className={`${
                  userId ? "" : "input-disabled cursor-not-allowed"
                }input-bordered mb-1 w-full rounded-sm border-[1px] p-2`}
              />
            </div>

            <div className="mb-[-10px] flex justify-end gap-2">
              <button
                className={`${
                  userId
                    ? " border-highlight-green bg-highlight-green "
                    : "input-disabled cursor-not-allowed bg-gray-400 text-black"
                } mt-[0.5rem] h-10 rounded-sm px-5`}
                onClick={handlePostButtonClick}
                disabled={userId ? false : true}
              >
                Post
              </button>
              <label
                className={`mt-[0.5rem] h-10 rounded-sm border-[2px] border-slate-200 bg-white px-5 pt-[0.38rem] text-center text-black hover:cursor-pointer`}
                htmlFor={`modal-${id}`}
              >
                Cancel
              </label>
            </div>
          </div>
        </label>
      </label>
    </>
  );
};
