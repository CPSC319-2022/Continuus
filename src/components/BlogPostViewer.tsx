import type { BlogPost, User, Comment } from "@prisma/client";
import { useState, useMemo, useEffect } from "react";
import { useInView } from "react-intersection-observer";
import { api } from "~/utils/api";
import { CommentModal } from "./CommentModal";
import { Spinner } from "./Spinner";
import { BlogPost as BlogPostComponent } from "./BlogPost";

type Post = BlogPost & { user: User; comments: (Comment & { user: User })[] };

export const BlogPostViewer: React.FC = () => {
  const [view, setView] = useState<string>("Recent");
  const {
    data: blogPosts,
    fetchNextPage,
    hasNextPage,
  } = api.blogPost.get.useInfiniteQuery(
    {
      take: 20,
    },
    { getNextPageParam: (lastPage) => ({ id: lastPage.nextCursor }) }
  );

  const posts = useMemo(
    () =>
      blogPosts?.pages.reduce(
        (result, curr) => [...result, ...curr.items],
        [] as Post[]
      ),
    [blogPosts]
  );

  const { ref, inView } = useInView();

  useEffect(() => {
    if (inView) {
      void (async () => {
        await fetchNextPage();
      })();
    }
  }, [inView, fetchNextPage]);

  return (
    <div className="w-full md:w-1/2 ">
      <div className="mb-6 flex w-full justify-end">
        <select
          className="h-8 w-64 max-w-xs border-b-2 bg-white"
          onChange={(e) => setView(e.target.value)}
        >
          <option>Recent</option>
          <option>Popular</option>
        </select>
      </div>
      {posts?.length === 0
        ? "Nothing to see here"
        : posts?.map(
            ({
              id,
              title,
              updatedAt,
              content,
              comments,
              user: { name, image },
            }) => (
              <>
                <div key={id} className="mb-6">
                  <BlogPostComponent
                    id={id}
                    name={name as string}
                    title={title}
                    lastUpdated={updatedAt}
                    imageUrl={image as string}
                    content={content}
                    comments={(comments as Comment[]).length}
                  />
                </div>
                <CommentModal
                  id={id}
                  title={title}
                  comments={comments as (Comment & { user: User })[]}
                  poster={name as string}
                  lastUpdated={updatedAt}
                  post={content}
                  posterAvatarUrl={image as string}
                />
              </>
            )
          )}
      {hasNextPage && (
        <div className="my-5 flex justify-center" ref={ref}>
          <Spinner size={2} />
        </div>
      )}
    </div>
  );
};
