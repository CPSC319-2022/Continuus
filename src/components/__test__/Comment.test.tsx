import { describe, it, expect, beforeEach, vi } from "vitest";
import { cleanup, render, screen } from "@testing-library/react";
import { CommentProps } from "~/components/Comment";
import userEvent from "@testing-library/user-event";
import { renderWithProviders } from "./utils";

const mockCommentProps: CommentProps = {
  dateAdded: new Date(),
  comment: "Hello world",
  name: "Test User",
  userId: "test-user-id",
  imgUrl: "https://i.stack.imgur.com/34AD2.jpg",
  commentId: "test-comment-id",
  dateUpdated: new Date(),
};

vi.mock("next-auth/react", () => ({
  useSession: () => ({
    data: {
      expires: "1",
      user: { id: "test-user-id", email: "a", name: "hi", image: "c" },
    },
    status: "authenticated",
  }),
}));

const initialMock = {
  api: {
    comment: {
      getUnique: {
        useQuery: () => ({
          data: {
            id: "test-comment-id",
            userId: "test-user-id",
            blogPostId: "test-blog-post-id",
            content: "Hello world",
            createdAt: new Date(),
            updatedAt: new Date(),
          },
          isLoading: false,
        }),
      },
      update: {
        useMutation: () => ({
          isSuccess: false,
        }),
      },
      delete: {
        useMutation: () => ({
          isSuccess: false,
        }),
      },
    },
    user: {
      currentUser: {
        useQuery: () => ({
          data: {
            createdAt: new Date(),
            email: "email",
            id: "test-user-id",
            image: "image",
            name: "contributor",
            role: "CONTRIBUTOR",
            contributorRequest: [],
            updatedAt: new Date(),
          },
          isLoading: false,
        }),
      },
    },
    useContext: () => ({
      comment: {
        get: {
          invalidate: vi.fn(),
        },
        count: {
          invalidate: vi.fn(),
        },
      },
    }),
  },
};

describe("Snapshot: Comment", () => {
  beforeEach(() => {
    cleanup();
    vi.clearAllMocks();
    vi.resetModules();
  });

  const setup = (mock: any) => {
    vi.doMock("~/utils/api", () => ({ ...mock }));
  };

  it("render comment", async () => {
    setup(initialMock);
    const { Comment } = await import("~/components/Comment");
    const body = renderWithProviders(
      <Comment {...mockCommentProps} />
    ).baseElement;
    expect(body).toMatchSnapshot();
  });

  it("render comment updated", async () => {
    setup({
      ...initialMock,
      api: {
        ...initialMock.api,
        comment: {
          ...initialMock.api.comment,
          update: { useMutation: () => ({ isSuccess: true }) },
        },
      },
    });
    const { Comment } = await import("~/components/Comment");
    const body = renderWithProviders(
      <Comment {...mockCommentProps} />
    ).baseElement;
    expect(body).toMatchSnapshot();
  });

  it("render comment deleted", async () => {
    setup({
      ...initialMock,
      api: {
        ...initialMock.api,
        comment: {
          ...initialMock.api.comment,
          delete: { useMutation: () => ({ isSuccess: true }) },
        },
      },
    });
    const { Comment } = await import("~/components/Comment");
    const body = renderWithProviders(
      <Comment {...mockCommentProps} />
    ).baseElement;
    expect(body).toMatchSnapshot();
  });

  it("editing comment", async () => {
    setup(initialMock);
    const { Comment } = await import("~/components/Comment");
    const body = renderWithProviders(<Comment {...mockCommentProps} />);
    const button = screen.getAllByTestId("comment-actions-button")[0];
    await userEvent.click(button!);
    const editButton = screen.getByTestId("edit-comment");
    await userEvent.click(editButton);
    expect(body.baseElement).toMatchSnapshot();
  });

  it("deleting comment", async () => {
    setup(initialMock);
    const { Comment } = await import("~/components/Comment");
    const body = renderWithProviders(<Comment {...mockCommentProps} />);
    const button = screen.getAllByTestId("comment-actions-button")[0];
    await userEvent.click(button!);
    const deleteButton = screen.getByTestId("delete-comment");
    await userEvent.click(deleteButton);
    expect(body.baseElement).toMatchSnapshot();
  });
});
