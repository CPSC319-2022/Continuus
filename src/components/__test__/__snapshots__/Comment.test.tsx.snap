// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot: Comment > deleting comment 1`] = `
<body>
  <div>
    <div
      class="mb-4"
    >
      <button
        class="absolute right-12 rounded-full transition-colors transition-all after:content-['\\\\2715'] hover:scale-125 hover:text-highlight-red"
      />
      <div
        class="flex h-20 w-full items-center justify-center"
      >
        <div
          class="text-lg"
        >
          Are you sure you want to delete this comment?
        </div>
      </div>
      <div
        class="flex justify-end border-b pb-3"
      >
        <button
          class="h-10 w-32 rounded-md border bg-highlight-red text-center text-white transition-all hover:cursor-pointer hover:bg-gray-700"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`Snapshot: Comment > editing comment 1`] = `
<body>
  <div>
    <div
      class="relative m-3 rounded-md border-b p-4"
    >
      <div
        class="flex"
      >
        <div
          class="avatar self-center transition-all hover:scale-110"
        >
           
          <a
            href="/profile/test-user-id"
          >
            <img
              alt="avatar"
              class="inline-block aspect-square rounded-[50%] "
              src="https://i.stack.imgur.com/34AD2.jpg"
              style="width: 2.5rem;"
            />
          </a>
        </div>
        <div
          class="ml-3"
        >
          <a
            href="/profile/test-user-id"
          >
            <p
              class="text-lg font-bold hover:underline"
            >
              Test User
            </p>
          </a>
          <p
            class="tex-gray-400 text-sm"
          >
            0 seconds ago
          </p>
        </div>
        <div
          class="absolute right-4 mt-[-12px] self-center"
        >
          <div
            class="dropdown-left dropdown rounded-md"
          >
            <button
              class="transition-all hover:scale-125"
              data-testid="comment-actions-button"
            >
              <div>
                <svg
                  height="24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 20q-.825 0-1.412-.587Q10 18.825 10 18q0-.825.588-1.413Q11.175 16 12 16t1.413.587Q14 17.175 14 18q0 .825-.587 1.413Q12.825 20 12 20Zm0-6q-.825 0-1.412-.588Q10 12.825 10 12t.588-1.413Q11.175 10 12 10t1.413.587Q14 11.175 14 12q0 .825-.587 1.412Q12.825 14 12 14Zm0-6q-.825 0-1.412-.588Q10 6.825 10 6t.588-1.412Q11.175 4 12 4t1.413.588Q14 5.175 14 6t-.587 1.412Q12.825 8 12 8Z"
                  />
                </svg>
              </div>
            </button>
            <ul
              class="dropdown-content menu mt-8 mr-[-16px] w-52 rounded-md border bg-white shadow"
              tabindex="0"
            >
              <li
                class="p-3 text-start transition-all hover:cursor-pointer hover:bg-gray-100"
                data-testid="edit-comment"
              >
                Edit
              </li>
              <li
                class="p-3 text-start transition-all hover:cursor-pointer hover:bg-gray-100"
                data-testid="delete-comment"
              >
                Delete
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="mt-3 p-2"
        data-testid="edit-comment-div"
      >
        <input
          class="input-bordered mb-3 w-full rounded-md border p-3"
          placeholder="Can't be empty!"
          type="text"
          value="Hello world"
        />
        <div
          class="flex justify-end gap-2"
        >
          <button
            class="h-10 w-32 rounded-md border-highlight-green bg-highlight-green transition-all hover:bg-gray-700 hover:text-highlight-green"
          >
            Save
          </button>
          <button
            class="h-10 w-32 rounded-md border bg-white text-center text-black transition-all hover:cursor-pointer hover:bg-highlight-red hover:text-white"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Snapshot: Comment > render comment 1`] = `
<body>
  <div>
    <div
      class="relative m-3 rounded-md border-b p-4"
    >
      <div
        class="flex"
      >
        <div
          class="avatar self-center transition-all hover:scale-110"
        >
           
          <a
            href="/profile/test-user-id"
          >
            <img
              alt="avatar"
              class="inline-block aspect-square rounded-[50%] "
              src="https://i.stack.imgur.com/34AD2.jpg"
              style="width: 2.5rem;"
            />
          </a>
        </div>
        <div
          class="ml-3"
        >
          <a
            href="/profile/test-user-id"
          >
            <p
              class="text-lg font-bold hover:underline"
            >
              Test User
            </p>
          </a>
          <p
            class="tex-gray-400 text-sm"
          >
            0 seconds ago
          </p>
        </div>
        <div
          class="absolute right-4 mt-[-12px] self-center"
        >
          <div
            class="dropdown-left dropdown rounded-md"
          >
            <button
              class="transition-all hover:scale-125"
              data-testid="comment-actions-button"
            >
              <div>
                <svg
                  height="24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 20q-.825 0-1.412-.587Q10 18.825 10 18q0-.825.588-1.413Q11.175 16 12 16t1.413.587Q14 17.175 14 18q0 .825-.587 1.413Q12.825 20 12 20Zm0-6q-.825 0-1.412-.588Q10 12.825 10 12t.588-1.413Q11.175 10 12 10t1.413.587Q14 11.175 14 12q0 .825-.587 1.412Q12.825 14 12 14Zm0-6q-.825 0-1.412-.588Q10 6.825 10 6t.588-1.412Q11.175 4 12 4t1.413.588Q14 5.175 14 6t-.587 1.412Q12.825 8 12 8Z"
                  />
                </svg>
              </div>
            </button>
            <ul
              class="dropdown-content menu mt-8 mr-[-16px] w-52 rounded-md border bg-white shadow"
              tabindex="0"
            >
              <li
                class="p-3 text-start transition-all hover:cursor-pointer hover:bg-gray-100"
                data-testid="edit-comment"
              >
                Edit
              </li>
              <li
                class="p-3 text-start transition-all hover:cursor-pointer hover:bg-gray-100"
                data-testid="delete-comment"
              >
                Delete
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="break-words"
      >
        <p
          class="mt-4 w-full"
        >
          Hello world
        </p>
      </div>
    </div>
  </div>
</body>
`;

exports[`Snapshot: Comment > render comment deleted 1`] = `
<body>
  <div />
</body>
`;

exports[`Snapshot: Comment > render comment updated 1`] = `
<body>
  <div>
    <div
      class="relative m-3 rounded-md border-b p-4"
    >
      <div
        class="flex"
      >
        <div
          class="avatar self-center transition-all hover:scale-110"
        >
           
          <a
            href="/profile/test-user-id"
          >
            <img
              alt="avatar"
              class="inline-block aspect-square rounded-[50%] "
              src="https://i.stack.imgur.com/34AD2.jpg"
              style="width: 2.5rem;"
            />
          </a>
        </div>
        <div
          class="ml-3"
        >
          <a
            href="/profile/test-user-id"
          >
            <p
              class="text-lg font-bold hover:underline"
            >
              Test User
            </p>
          </a>
          <p
            class="tex-gray-400 text-sm"
          >
            0 seconds ago
          </p>
        </div>
        <div
          class="absolute right-4 mt-[-12px] self-center"
        >
          <div
            class="dropdown-left dropdown rounded-md"
          >
            <button
              class="transition-all hover:scale-125"
              data-testid="comment-actions-button"
            >
              <div>
                <svg
                  height="24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 20q-.825 0-1.412-.587Q10 18.825 10 18q0-.825.588-1.413Q11.175 16 12 16t1.413.587Q14 17.175 14 18q0 .825-.587 1.413Q12.825 20 12 20Zm0-6q-.825 0-1.412-.588Q10 12.825 10 12t.588-1.413Q11.175 10 12 10t1.413.587Q14 11.175 14 12q0 .825-.587 1.412Q12.825 14 12 14Zm0-6q-.825 0-1.412-.588Q10 6.825 10 6t.588-1.412Q11.175 4 12 4t1.413.588Q14 5.175 14 6t-.587 1.412Q12.825 8 12 8Z"
                  />
                </svg>
              </div>
            </button>
            <ul
              class="dropdown-content menu mt-8 mr-[-16px] w-52 rounded-md border bg-white shadow"
              tabindex="0"
            >
              <li
                class="p-3 text-start transition-all hover:cursor-pointer hover:bg-gray-100"
                data-testid="edit-comment"
              >
                Edit
              </li>
              <li
                class="p-3 text-start transition-all hover:cursor-pointer hover:bg-gray-100"
                data-testid="delete-comment"
              >
                Delete
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="break-words"
      >
        <p
          class="mt-4 w-full"
        >
          Hello world
        </p>
      </div>
    </div>
  </div>
</body>
`;
